spannerlib/grpc-server/server.go

```diff
@@ -127,38 +127,39 @@ func (s *spannerLibServer) ExecuteStreaming(request *pb.ExecuteRequest, stream g
 	if err != nil {
 		return err
 	}
-	defer func() { _ = api.CloseRows(context.Background(), request.Connection.Pool.Id, request.Connection.Id, id) }()
 	rows := &pb.Rows{Connection: request.Connection, Id: id}
-	metadata, err := api.Metadata(queryContext, request.Connection.Pool.Id, request.Connection.Id, id)
+	return s.streamRows(queryContext, rows, stream)
+}
+
+func (s *spannerLibServer) streamRows(queryContext context.Context, rows *pb.Rows, stream grpc.ServerStreamingServer[pb.RowData]) error {
+	defer func() { _ = api.CloseRows(context.Background(), rows.Connection.Pool.Id, rows.Connection.Id, rows.Id) }()
+	metadata, err := api.Metadata(queryContext, rows.Connection.Pool.Id, rows.Connection.Id, rows.Id)
 	if err != nil {
 		return err
 	}
 
 	first := true
 	for {
-		if queryContext.Err() != nil {
-			return queryContext.Err()
-		}
-		if row, err := api.Next(queryContext, request.Connection.Pool.Id, request.Connection.Id, id); err != nil {
+		if row, err := api.Next(queryContext, rows.Connection.Pool.Id, rows.Connection.Id, rows.Id); err != nil {
 			return err
 		} else {
 			if row == nil {
-				stats, err := api.ResultSetStats(queryContext, request.Connection.Pool.Id, request.Connection.Id, id)
+				stats, err := api.ResultSetStats(queryContext, rows.Connection.Pool.Id, rows.Connection.Id, rows.Id)
 				if err != nil {
 					return err
 				}
-				nextMetadata, err := api.NextResultSet(queryContext, request.Connection.Pool.Id, request.Connection.Id, id)
-				if err != nil {
-					return err
-				}
-				res := &pb.RowData{Rows: rows, Stats: stats, HasMoreResults: nextMetadata != nil}
+				nextMetadata, nextResultSetErr := api.NextResultSet(queryContext, rows.Connection.Pool.Id, rows.Connection.Id, rows.Id)
+				res := &pb.RowData{Rows: rows, Stats: stats, HasMoreResults: nextMetadata != nil || nextResultSetErr != nil}
 				if first {
 					res.Metadata = metadata
 					first = false
 				}
 				if err := stream.Send(res); err != nil {
 					return err
 				}
+				if nextResultSetErr != nil {
```

Comment:
This ensures that calling `NextResultSet` when using the streaming gRPC API returns an error if the second (or third, ...) SQL statement in a multi-statement SQL string fails.